test/decl/protocol/conforms/self.swift

```diff
@@ -86,3 +86,19 @@ extension Node {
 }
 
 class IntNode: Node {}
+
+// SR-8902
+protocol P8902 {
+    associatedtype A // expected-note {{protocol requires nested type 'A'; do you want to add it?}}
+    func f(_ x: A) -> Self
+}
+struct S : P8902 {
+    func f(_ x: Bool) -> S { fatalError() }
+}
+class C8902 : P8902 { // expected-error {{type 'C8902' does not conform to protocol 'P8902'}}
```

Comment:
Can you add a non-final class where we have to infer A but the witness returns Self?

Reply:
Yep! You were right, needed an extra check for that.